update-testsuite.py

```diff
@@ -0,0 +1,203 @@
+    spec3 = Repo('spec', branch='wasm-3.0')
+
+    repos = []
+    repos.append(spec3)
+    repos.append(Repo('threads'))
+    repos.append(Repo('exception-handling'))
+    repos.append(Repo('gc'))
+    repos.append(Repo('tail-call'))
+    repos.append(Repo('annotations'))
+    repos.append(Repo('function-references'))
+    repos.append(Repo('memory64', upstream=spec3))
+    repos.append(Repo('extended-const'))
+    repos.append(Repo('multi-memory'))
+    repos.append(Repo('relaxed-simd'))
+    repos.append(Repo('custom-page-sizes'))
+    repos.append(Repo('wide-arithmetic'))
+
+    # Make sure that `repos/spec` is a git repository
+    if not os.path.isdir('repos/spec'):
+        git('init', 'repos/spec')
+
+    # Update the spec itself, reset to the latest version of the spec, and then
+    # copy all files from the upstream spec tests into this repository's own
+    # suite of tests to run.
+    spec.update()
+    spec.checkout_merge()
+    for path, _repo_path, dst in spec.list_tests():
+        shutil.copyfile(path, dst)
+
+    # Process all upstream repositories and proposals.
+    failed_merges = []
+    updated = []
+    for repo in repos:
+        # Repositories may not be mergable with the upstream spec repository in
+        # which case we skip them and print an informational message at the end.
+        repo.update()
+        if not repo.checkout_merge(spec):
+            failed_merges.append(repo)
+            continue
+
+        # Blow away this proposal's list of tests if it exists.
+        dstdir = f'proposals/{repo.dir}'
+        if os.path.isdir(dstdir):
+            shutil.rmtree(dstdir)
+
+        # For all tests in this proposal run a diff against the upstream
+        # revision. If the diff is non-empty then include this test by copying
+        # it to its destination.
+        for path, repo_path, dst in spec.list_tests():
+            upstream = repo.upstream or spec
+
+            diff = repo.git('diff', upstream.rev, repo.merged_rev, '--', repo_path, quiet=True)
+            if len(diff) == 0:
+                continue
+
+            dst = os.path.join(dstdir, dst)
+            os.makedirs(os.path.dirname(dst), exist_ok=True)
+            shutil.copyfile(path, dst)
+
+        # If anything changed, then this is an updated proposal, and take note
+        # of that for later.
+        git('add', dstdir)
+        status = git('status', '-s', dstdir)
+        if len(status) > 0:
+            updated.append(repo)
+
+    for repo in failed_merges:
+        print('!! failed to update:', repo.url())
+
+    # If anything was updated, make a commit message indicating as such.
+    if len(updated) == 0:
+        return
+    message = "Update repos:\n\n"
+    for repo in updated:
+        message += f"  {repo.dir}:\n"
+        message += f"    {repo.url()}/commit/{repo.rev}\n"
+    message += "\n"
+    message += "This change was automatically generated by `update-testsuite.py`"
+    git('commit', '-a', '-m', message)
+
```

Comment:
Python style (at least flake8) is to have two empty lines between top level functions/classes.